File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def findMinimum(self,nums:List[int]):
        n = len(nums)
        left = 0
        right = n-1

        while(left<right):
            mid = left+(right-left)//2
            if nums[mid]>nums[right]:
                left = mid + 1
            else:
                right = mid

        return right    

    def BinarySearch(self,left,right,target,nums):
        while(left<=right):
            mid = left+(right-left)//2
            if nums[mid]==target:
                return mid
            elif nums[mid]<target:
                left = mid + 1
            else:
                right = mid - 1

        return -1                        


    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        pivot_index = self.findMinimum(nums)

        idx = self.BinarySearch(0,pivot_index-1,target,nums)

        if idx != -1:
            return idx

        idx = self.BinarySearch(pivot_index,n-1,target,nums)
        return idx 

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int solveFindMinBS(vector<int>& nums){
        int s = 0;
        int e = nums.size() - 1;
        while(s<e){
            int mid = s+(e-s)/2;

            if(nums[mid]>nums[e]) s = mid + 1;
            else e = mid;
        }
        return e;
    }
    int binarySearch(vector<int>& nums, int target, int s , int e){
        while(s<=e){
            int mid = s+(e-s)/2;
            if(nums[mid]==target) return mid;
            else if(nums[mid]<target){
                s = mid + 1;
            }else{
                e = mid - 1;
            }
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int pivot = solveFindMinBS(nums);
        int leftSortedAns = binarySearch(nums,target,0,pivot-1);
        int rightSortedAns = binarySearch(nums,target,pivot,nums.size()-1);
        if(leftSortedAns!= -1) return leftSortedAns;
        return rightSortedAns;
    }
};